File: plugins/ai-pilot/scripts/lib/apex/rollout-locate.ts
/**
 * rollout-locate.ts - Locate the Codex session's rollout JSONLs (main + children).
 *
 * Codex runs APEX research in spawned subagents, whose MCP calls land in CHILD
 * rollout files (own thread UUID) that record the parent via parent_thread_id /
 * the shared session_id (openai/codex#20437). A gate must therefore scan the
 * main rollout AND its children, not just the file named after session_id.
 * Per-tool PostToolUse is unreliable in code_mode (openai/codex#19385), so the
 * rollout tree is the ground truth.
 */
import { readdirSync, statSync, openSync, readSync, closeSync } from "node:fs";

const CODEX_HOME = process.env.CODEX_HOME ?? `${process.env.HOME}/.codex`;
const MAX_TAIL_BYTES = 8 * 1024 * 1024;  // memory backstop; real scoping is turn-anchoring below
const MAX_AGE_MS = 3_600_000;
const HEAD_BYTES = 4096;
// Current-turn anchors: match compact (Codex/Rust) and spaced (Python) JSON.
const USER_TURN_MARKERS = ['"type": "user_message"', '"type":"user_message"'];
const TASK_START_MARKERS = ['"type": "task_started"', '"type":"task_started"'];

/** Recursively collect recent *.jsonl rollout paths (bounded by mtime). */
function recentRollouts(dir: string, depth: number, out: string[]): void {
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return;
  }
  for (const e of entries) {
    const path = `${dir}/${e}`;
    let st;
    try {
      st = statSync(path);
    } catch {
      continue;
    }
    if (st.isDirectory()) {
      if (depth > 0) recentRollouts(path, depth - 1, out);
    } else if (e.endsWith(".jsonl") && Date.now() - st.mtimeMs <= MAX_AGE_MS) {
      out.push(path);
    }
  }
}

/** Read the first HEAD_BYTES of a file (session_meta lives on line 1). */
function readChunk(path: string, fromEnd: boolean): string {
  let fd = -1;
  try {
    fd = openSync(path, "r");
    const size = statSync(path).size;
    const length = Math.min(fromEnd ? MAX_TAIL_BYTES : HEAD_BYTES, size);
    const start = fromEnd ? size - length : 0;
    const buf = Buffer.alloc(length);
    readSync(fd, buf, 0, length, start);
    return buf.toString("utf-8");
  } catch {
    return "";
  } finally {
    if (fd >= 0) try { closeSync(fd); } catch { /* noop */ }
  }
}

/** Slice `text` from the current task turn: the last user_message line, else
 * the last task_started (subagent child rollouts have no user_message), else
 * the whole tail. Scoping by turn — not a fixed byte window — stops a noisy
 * command from evicting the turn's evidence. */
function sliceFromTurnAnchor(text: string): string {
  for (const markers of [USER_TURN_MARKERS, TASK_START_MARKERS]) {
    const at = Math.max(...markers.map((m) => text.lastIndexOf(m)));
    if (at >= 0) {
      const lineStart = text.lastIndexOf("\n", at);
      return text.slice(lineStart >= 0 ? lineStart + 1 : 0);
    }
  }
  return text;
}

/** Read the MAX_TAIL_BYTES tail, sliced to the current task turn. */
export function readTail(path: string): string {
  return sliceFromTurnAnchor(readChunk(path, true));
}

/**
 * Return the session's rollout files: the one named after session_id plus any
 * child rollout whose session_meta references session_id (parent linkage).
 */
export function sessionRollouts(sessionId: string): string[] {
  if (!sessionId) return [];
  const all: string[] = [];
  recentRollouts(`${CODEX_HOME}/sessions`, 4, all);
  const out: string[] = [];
  for (const path of all) {
    const name = path.slice(path.lastIndexOf("/") + 1);
    if (name.includes(sessionId) || readChunk(path, false).includes(sessionId)) {
      out.push(path);
    }
  }
  return out;
}

File: plugins/_shared/scripts/rollout_locate.py
#!/usr/bin/env python3
"""Locate/parse the Codex session rollout tree (main + child subagent rollouts).

PostToolUse is unreliable in code_mode (openai/codex#19385) and research runs in
subagents (own rollouts, linked via session_id), so scan the whole tree.
"""
import glob
import json
import os
import time

CODEX_HOME = os.environ.get("CODEX_HOME", os.path.join(os.path.expanduser("~"), ".codex"))
MAX_TAIL_BYTES = 8 * 1024 * 1024  # memory backstop; real scoping is turn-anchoring below
MAX_AGE_SECONDS = 3600
HEAD_BYTES = 4096
# Current-turn anchors: last user_message, else task_started (subagent child
# rollouts have none). Match compact (Codex/Rust) and spaced (Python) JSON.
TURN_MARKERS = (('"type": "user_message"', '"type":"user_message"'),
                ('"type": "task_started"', '"type":"task_started"'))


def _recent_rollouts() -> list:
    """Return rollout paths modified within MAX_AGE_SECONDS (bounded scan)."""
    pattern = os.path.join(CODEX_HOME, "sessions", "**", "*.jsonl")
    now = time.time()
    out = []
    for path in glob.glob(pattern, recursive=True):
        try:
            if now - os.path.getmtime(path) <= MAX_AGE_SECONDS:
                out.append(path)
        except OSError:
            continue
    return out


def _head(path: str) -> str:
    """Return the first HEAD_BYTES of a file (session_meta lives on line 1)."""
    try:
        with open(path, "rb") as fh:
            return fh.read(HEAD_BYTES).decode("utf-8", "replace")
    except OSError:
        return ""


def session_rollouts(session_id: str) -> list:
    """Main rollout (filename UUID) plus children referencing session_id."""
    if not session_id:
        return []
    out = []
    for path in _recent_rollouts():
        if session_id in os.path.basename(path) or session_id in _head(path):
            out.append(path)
    return out


def _anchor_index(lines: list) -> int:
    """Index of the current turn's start (last user_message, else task_started)."""
    for markers in TURN_MARKERS:
        for i in range(len(lines) - 1, -1, -1):
            if any(m in lines[i] for m in markers):
                return i
    return 0


def tail_lines(path: str) -> list:
    """Current turn's lines: MAX_TAIL_BYTES tail sliced from the anchor."""
    try:
        size = os.path.getsize(path)
        with open(path, "rb") as fh:
            if size > MAX_TAIL_BYTES:
                fh.seek(size - MAX_TAIL_BYTES)
            data = fh.read()
    except OSError:
        return []
    lines = data.decode("utf-8", "replace").splitlines()
    return lines[_anchor_index(lines):]


def _payload(entry: dict) -> dict:
    """Normalise a rollout entry to its function_call payload across formats."""
    payload = entry.get("payload")
    if isinstance(payload, dict):
        return payload
    item = entry.get("item")
    return item if isinstance(item, dict) else entry


def iter_function_calls(session_id: str, ttl_seconds: int = 0):  # noqa: ARG001
    """Yield CURRENT-TURN function_call payloads (tail_lines is turn-scoped; ttl ignored)."""
    for path in session_rollouts(session_id):
        for line in tail_lines(path):
            if "function_call" not in line:
                continue
            try:
                entry = json.loads(line)
            except (json.JSONDecodeError, ValueError):
                continue
            payload = _payload(entry)
            if payload.get("type") == "function_call":
                yield payload
